signer/src/stacks/contracts.rs

```diff
@@ -987,11 +1066,118 @@ impl AsContractCall for RejectWithdrawalV1 {
+            return Err(WithdrawalRejectErrorMsg::RequestMissing.into_error(req_ctx, self));
+        };
+
+        eprintln!("check withdrawal request report status");
+        match report.status {
+            WithdrawalRequestStatus::Confirmed => (),
+            WithdrawalRequestStatus::Fulfilled(_txid) => {
+                // fails #2
+                eprintln!("withdrawal request was fulfilled");
+                return Err(
+                    WithdrawalRejectErrorMsg::WithdrawalRequestFulfilled.into_error(req_ctx, self)
+                );
+            }
+            WithdrawalRequestStatus::Unconfirmed => {
+                // fails #1
+                eprintln!("withdrawal request was unconfirmed");
+                return Err(WithdrawalRejectErrorMsg::WithdrawalRequestUnconfirmed
+                    .into_error(req_ctx, self));
+            }
+        }
+
+        eprintln!("validate final request");
+        // 3. The request is final. Check that we've seen six new bitcoin blocks since observing the bitcoin anchor block associated with the Stacks block confirming the withdrawal request transaction. Fail the withdrawal request if we’ve observed less than 6 bitcoin blocks since the anchor block.
+
+        let request_block_height = report.block_height;
+
+        let Some(tip_block_hash) = ctx.get_storage().get_bitcoin_canonical_chain_tip().await?
+        else {
+            eprintln!("no bitcoin canonical chain tip");
+            return Err(WithdrawalRejectErrorMsg::NoBitcoinChainTip.into_error(req_ctx, self));
+        };
+
+        let Some(tip_bitcoin_block) = ctx.get_storage().get_bitcoin_block(&tip_block_hash).await?
+        else {
+            eprintln!("no bitcoin block for canonical chain tip");
+            return Err(WithdrawalRejectErrorMsg::MissingBitcoinBlock.into_error(req_ctx, self));
+        };
+
+        let tip_block_height = tip_bitcoin_block.block_height;
+        if tip_block_height < (request_block_height + 6) {
+            return Err(
+                WithdrawalRejectErrorMsg::WithdrawalRequestNotFinal.into_error(req_ctx, self)
+            );
+        }
+
+        eprintln!("validate bitmap");
+        // 4. The request is collectively rejected or expired.
+        eprintln!("get withdrawal request signer votes");
+        let signer_votes = ctx
+            .get_storage()
+            .get_withdrawal_request_signer_votes(&qualified_id, &req_ctx.aggregate_key)
+            .await
+            .map_err(|_| WithdrawalRejectErrorMsg::RequestMissing.into_error(req_ctx, self))?;
+        let signer_bitmap = BitArray::<[u8; 16]>::from(signer_votes);
+        eprintln!("signer_votes  {}", &signer_bitmap);
+        eprintln!("signer_bitmap {}", &self.signer_bitmap);
+
+        if signer_bitmap != self.signer_bitmap {
+            return Err(WithdrawalRejectErrorMsg::BitmapMismatch.into_error(req_ctx, self));
+        }
+
+        // TODO: how many votes are needed to be valid?
```

Comment:
I think you need `num_signers - threshold + 1` votes against in order to know that the withdrawal request will be fulfilled. And we should use the `withdrawal_signers` variable above for this, since those are the raw actual votes; `get_withdrawal_request_signer_votes` massages missing votes into votes against, which we don't want.

Reply:
Should we still validate that the bitmap from `get_withdrawal_request_signer_votes` matches `self.signer_bitmap` like `AcceptWithdrawalV1` does?

Reply:
Yes, we should. We need to do something for the votes that we are missing, so using `get_withdrawal_request_signer_votes` for creating the bitmap makes sense. But for making sure that we have enough actual votes against, we need to use that other query, the one in `get_withdrawal_signers`.

Reply:
Ok, I'm now iterating over the withdrawal signers returned from `get_withdrawal_signers`, and summing up the `is_accepted` count.   Looks like I can get the threshold via ```
  Some(latest_dkg) = db.get_latest_encrypted_dkg_shares
  latest_dkg.signature_share_threshold
```

Reply:
Ok, I fixed the withdrawal request block height in `TestSweepSetup2`, it was using `Fake` data 😹 

I also hardwired the correct `threshold` into the `TestSweetSetup2` version of `ReqContext`. 

Now I just have to tell the test harness to vote against the withdrawal request, and this should be in a much better place.